source/extensions/filters/http/api_key_auth/api_key_auth.cc

```diff
@@ -135,6 +167,22 @@ Http::FilterHeadersStatus ApiKeyAuthFilter::decodeHeaders(Http::RequestHeaderMap
     }
   }
 
+  if (forwarding.has_value()) {
+    const Http::LowerCaseString& header_name = forwarding->getHeaderName();
+
+    if (!header_name.get().empty()) {
+      const std::string& client = credential->second;
+      // Add the authenticated client to the request headers.
+      headers.setCopy(header_name, client);
```

Comment:
`SetReferenceKey` because the header key will always be alive untill the request is end.